GSA-TTS/FAC
# General Information - Allowed to be Partially Complete

# General Information - Allowed to be Partially Complete

Issue: https://github.com/GSA-TTS/FAC/issues/1992

## Changes:
1. The general information form can now be submitted partially complete.
    * HTML reflects this difference in that nothing is required anymore, and many error messages were struck.
    * JS reflects this difference in that no fields are made required, and null value errors have been removed.
    * The django form no longer requires all the fields, and has some regex validators to limit 400 responses from the gen info page
    * The validator (jsonnet schema) still runs, but it allows most fields to be empty.

2. The submission checklist page runs a different validator on the general information section
    * The new schema is `GeneralInformationComplete.jsonnet`, and it's similar to the other one, but it allows no empty data and requires all necessary fields.
    * When visiting the submission checklist page, you'll be able to tell when the form hasn't been totally completed.

3. Many test changes:
    * Some now need to be looking for errors in the page, rather than 400 responses
    * Some need to handle dates a little differently
    
4. Some other things:
    * The sidenav (top left) is fixed. It now has three sections that match up with the three major form groups. That broke at some point as we were updating the form. 

## How to test:
1. Switch to this branch and start a clean run. This will clear old submissions, which may or may not now be broken. 
2. Start a new submission, get to the general information form.
3. Verify that:
    * You can exit at this point and see your submission in the homepage
    * Visiting the checklist for that submission displays the form as incomplete
    * Navigating to the general information form still autocomplete some sections (fiscal year dates, UEI)
4. Fill out some fields and submit the form. Verify that:
    * The form still shows as incomplete
    * Those fields saved properly
    * Those fields auto-fill when revisiting the form.
5. Fill out some fields very wrong on purpose. Try to submit, and verify that these fields give reasonable error messages:
    * EINs (non-numbers, too long/short values)
    * Ciities (numbers, special characters like underscores)
    * Dates (not MM/DD/YYYY)
    * Phone numbers
6. Finish the form, submit, and verify that it displays as complete.
7. Play around with submitting/not submitting certain fields to try and find a case where it shows as complete when it isn't, or the opposite. 
8. Step through to cross-al and verify that general information gives errors if it isn't done, and gives no errors if it is done.

## Screenshots:
Submission progress page, incomplete general information:
![image](https://github.com/GSA-TTS/FAC/assets/91098850/175359b6-e2ea-42d1-b765-1cd2effbb84e)

Submission progress page, completed general information (the same as before):
![image](https://github.com/GSA-TTS/FAC/assets/91098850/7e2e4121-8f0a-4393-95bf-373db515a3bc)


General information with errors:
![image](https://github.com/GSA-TTS/FAC/assets/91098850/47c27ced-c467-4f39-9cc2-6bac564cb818)



## PR checklist: submitters


## PR checklist: reviewers

        
The larger the PR, the stricter we should be about these points.
